lib/ui/painting.dart

```diff
@@ -3464,6 +3628,121 @@ class MaskFilter {
   String toString() => 'MaskFilter.blur($_style, ${_sigma.toStringAsFixed(1)})';
 }
 
+abstract class _ColorTransform {
+  Color transform(Color color, ColorSpace resultColorSpace);
+}
+
+class _IdentityColorTransform implements _ColorTransform {
+  const _IdentityColorTransform();
+  @override
+  Color transform(Color color, ColorSpace resultColorSpace) => color;
+}
+
+class _ClampTransform implements _ColorTransform {
+  const _ClampTransform(this.child);
+  final _ColorTransform child;
+  @override
+  Color transform(Color color, ColorSpace resultColorSpace) {
+    return Color.from(
+      alpha: clampDouble(color.a, 0, 1),
+      red: clampDouble(color.r, 0, 1),
+      green: clampDouble(color.g, 0, 1),
+      blue: clampDouble(color.b, 0, 1),
+      colorSpace: resultColorSpace);
+  }
+}
+
+class _MatrixColorTransform implements _ColorTransform {
+  /// Row-major.
+  const _MatrixColorTransform(this.values);
+
+  final List<double> values;
+
+  @override
+  Color transform(Color color, ColorSpace resultColorSpace) {
+    return Color.from(
+        alpha: color.a,
+        red: values[0] * color.r +
+            values[1] * color.g +
+            values[2] * color.b +
+            values[3],
+        green: values[4] * color.r +
+            values[5] * color.g +
+            values[6] * color.b +
+            values[7],
+        blue: values[8] * color.r +
+            values[9] * color.g +
+            values[10] * color.b +
+            values[11],
+        colorSpace: resultColorSpace);
+  }
+}
+
+_ColorTransform _getColorTransform(ColorSpace source, ColorSpace destination) {
+  // The transforms were calculated with the following octave script from known
+  // conversions. These transforms have a white point that matches Apple's.
+  //
+  // p3Colors = [
+  //   1, 0, 0, 0.25;
+  //   0, 1, 0, 0.5;
+  //   0, 0, 1, 0.75;
+  //   1, 1, 1, 1;
+  // ];
+  // srgbColors = [
+  //   1.0930908918380737,  -0.5116420984268188, -0.0003518527664709836, 0.12397786229848862;
+  //   -0.22684034705162048, 1.0182716846466064,  0.00027732315356843174,  0.5073589086532593;
+  //   -0.15007957816123962, -0.31062406301498413, 1.0420056581497192,  0.771118700504303;
+  //   1,       1,       1,       1;
+  // ];
+  //
+  // format long
+  // p3ToSrgb = srgbColors * inv(p3Colors)
+  // srgbToP3 = inv(p3ToSrgb)
+  const _MatrixColorTransform srgbToP3 = _MatrixColorTransform(<double>[
```

Comment:
Can these values be const?

Reply:
Yep!  I wanted to make sure no garbage is generated when performing transforms.